### ISSUE TITLE
"Cannot change to" directory\n warning when entering directory ending with newline

### ISSUE THREAD

**Issue by mc-butler** [MidnightCommander/mc#2325]
> [!IMPORTANT]
> This issue was migrated from Trac:
>
> | **Origin** | https://midnight-commander.org/ticket/2325 |
> | :------------- | :------------------------------------------------------------------------------------------------ |
> | **Reporter** | `gotar` (gotar@….pl) |

<ol><li>create directory with trailing newline,
</li><li>get into it.
</li></ol><p>
The warning appears, although nothing unexpected happens and works as expected.<br />
</p>

**Comment by egmontkob** [comment]
`subshell_name_quote()` produces a string that looks like

    cd "`printf '%b' 'ABC\0nnnDEF\0nnnXYZ'`"

The problem is that ``` `...` ``` a.k.a. `$(...)` command substitution strips off all trailing newlines. So it's not a binary safe method.

Quite a few shells (at least bash, zsh and tcsh) support `$''` string definitions with escape characters, e.g.

    cd $'abc\n\n'

which is also much simpler. We should do this on at least those shells that support this (or some similar syntax for this).

**Comment by egmontkob** [comment]
[fish](https://stackoverflow.com/questions/59886560/how-can-i-have-a-newline-in-a-string-in-fish-shell):

    cd 'abc'\n\n

or any `\nnn` (octal) or `\xnn` (hex) outside of quote marks / apostophes.

**Comment by zyv** [comment]
@mirabilos, maybe with your knowledge of shells, you could help us out here: off the top of your head, do you know the equivalent of `cd $'abc\n\n'` (`bash` / `zsh` / `tcsh`) for our supported subshells receiving least love like `*ksh`, `ash`, `dash`, and friends?

**Comment by egmontkob** [comment]
The following would almost certainly be a **wrong** solution. Let me mention it before someone goes with it and doesn't see the downsides :)

Simply have literal newlines between single quotes. E.g. to enter the directory `abc\n\n`, it would be something like

```
cd 'abc

'
```

Why is it wrong?

We're targeting plain / simple / old-fashioned shells, and they don't have line editing, so we're entering these to the kernel's cooked mode. Depending on platform and stty settings, it might perform weird cr/lf conversions. Other special bytes (like ^C interrupt, ^U yank etc.) couldn't work this way either.

**Comment by egmontkob** [comment]
dash, ksh93, mksh:

I can't see how to specify string constants with embedded special characters in `dash` (except for a literally typing them, which we cannot do as per the previous comment).

What I've found though is that these shells support `${parameter%word}` to remove a suffix from a variable. Is this standard POSIX sh, supported by other shells too?

So what we could do: use `printf` in command substitution to print the working directory, but with an additional character (`_` in this example). This prevents chopping off trailing newlines. Intermediate newlines are preserved as expected. E.g. to enter the directory `abc\n\n`:

    newdir_="$(printf '%b_' 'abc\n\n')"

Then strip off that trailing `_`. It works because _here_ we don't use command substitution, but rather string operations which are fine with keeping a trailing newline:

    newdir="${newdir_%_}"
    cd "$newdir"

The last two could obviously be combined into a single statement.

How portable would this be?

**Comment by egmontkob** [comment]
tcsh:

Command substition also mangles _internal_ newlines, not just final ones. So this workaround cannot work. (Plus I haven't found under 1 minute how to chop off the last character.) Yup, new bugreport: currently with tcsh you can't enter directories with newlines anywhere in their name.

However, the much more straightforward `cd $'abc\n\n'`-style string literals do work there. And once we _have to_ implement it, bash and zsh and friends (whoever supports it) could also use it rather than the more convoluted printf-approach.

Looks like (if we're lucky) we can do with 3 different codes:
- straightforward string literal for bash, zsh, tcsh, and maybe more;
- slightly different straightforward string literal syntax for fish;
- printf, command substitution, stripping of a trailing character for old-fashioned sh (if they all support it) (or a better solution if we find one).

---

Let's not forget to address #4480 at the same time (it would be pointless to rewrite the code twice). We should make sure, for all these shells, never to emit a physical line longer than 256 chars; for each shell we should figure out how to wrap long logical lines into multiple physical ones.

**Comment by zyv** [comment]
> How portable would this be?

Wow, this reminds me of the old good autotools style `if test x$SHELL_VAR = xyes; then` tests... and is probably as simple and as portable as it can be.

**Comment by egmontkob** [comment]
`${parameter%word}` [is indeed POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_02), yay!

**Comment by ossilator** [comment]
> `${parameter%word}` is indeed POSIX, yay!

that doesn't mean that it would work on an old solaris. it probably would (can't remember problems with this), but that shell is/was a notorious PITA.

**Comment by ossilator** [comment]
> Let's not forget to address #4480 at the same time (it would be pointless to rewrite the code twice).

nah, this is a case where i'd actually prefer the churn, because the changes are both functional and logically rather separate. i'd make the line end change first.

**Comment by egmontkob** [comment]
> nah, this is a case where i'd actually prefer the churn, because the changes are both functional and logically rather separate.

With all due respect: I'm not gonna do this.

When at least 1, but rather both modifications are quite heavy, requiring an almost complete rewrite of a tiny portion of mc, it's just a pointless waste of time. Noticeably more coding work, double testing work, just to thoroughly test an intermediate version that won't be used by anyone, and _might_ save us a little time if the code turns out buggy and needs followup investigation. Even if we're only talking about half an hour - hour -ish amount of work.

It's quite a different story from the PS1 and cosmetical changes where it was trivial work; moreover only the code lines conflicted, the functionality didn't; and therefore I was happy to do it (I'm yet to continue that PR and upload these changes).

By the way, my first draft, addressing both issues, attached at https://github.com/MidnightCommander/mc/issues/4480#issuecomment-3401372437, was almost ready when you made this comment.

zyv already expressed that he's somewhat loose on these things, and prefers more effective use of our engineering time.

I'd also be happy to comply if my boss asked this, and paid for my time :)

If you're happy to work backwards from my (yet-to-be-cleaned-up) patch and create the (well tested) intermediate state, I'd be happy to have that committed before mine, sharing credits.

**Comment by ossilator** [comment]
my thinking was that the lineending change was already there, and rewriting it before proceeding with the control char injection wouldn't add extra work. i wouldn't test the intermediate state, because bugs in the parts that won't subsequently change would be caught in the final state, and could be fixed in the former commit with only minor inconvenience. going beyond that would indeed be a questionable trade-off under the given circumstances.
whatever ...
